chasm/tree.go

```diff
@@ -1262,30 +1307,91 @@ func (n *Node) closeTransactionHandleRootLifecycleChange() (bool, error) {
 }
 
 func (n *Node) closeTransactionForceUpdateVisibility(
-	nextVersionedTransition *persistencespb.VersionedTransition,
+	rootLifecycleChanged bool,
 ) error {
+	var visibilityNode *Node
 	for _, child := range n.children {
 		componentAttr := child.serializedNode.Metadata.GetComponentAttributes()
-		if componentAttr == nil ||
-			componentAttr.Type != visibilityComponentFqType ||
-			transitionhistory.Compare(child.serializedNode.Metadata.LastUpdateVersionedTransition, nextVersionedTransition) == 0 {
-			// Skip force update is the node is not a visibility component or already updated in this transition.
+		if componentAttr == nil {
 			continue
 		}
 
-		mutableContext := NewMutableContext(context.Background(), n)
-		visComponent, err := child.Component(mutableContext, ComponentRef{})
-		if err != nil {
-			return err
+		if child.valueState == valueStateNeedSerialize {
+			if rc, ok := n.registry.componentFor(child.value); ok && rc.fqType() == visibilityComponentFqType {
+				visibilityNode = child
+				break
+			}
+		} else if componentAttr.Type == visibilityComponentFqType {
+			visibilityNode = child
+			break
 		}
 
-		visibility, ok := visComponent.(*Visibility)
-		if !ok {
-			return serviceerror.NewInternalf("expected visibility component for component with type %s, but got %T", visibilityComponentFqType, visComponent)
+	}
+	if visibilityNode == nil {
+		return nil
+	}
+
+	needUpdate := rootLifecycleChanged
+
+	mutableContext := NewMutableContext(context.Background(), n)
```

Comment:
Is there a good reason to construct a `context.Context` object every time here? You can just pass in nil or a static context object.
Looks like the only reason we hold a context object is to pass in an intent.
I also don't see a reason for a library author to ever create context objects directly but can't come up with a way to hide this method since the non-chasm part of the codebase needs it.
I'll need to think about it some more, it's late.

Reply:
yeah application code has no way to get and use that context.  
It's more like a placeholder for later partial read work where the framework potentially needs to go to persistence and load some component states. Let me actually change it to context.TODO() here and add comments.